### PR TITLE
[ConstantFold] Drop gep of gep fold entirely

### PR DESCRIPTION
**Pull request by nikic**
This is a followup to https://github.com/llvm/llvm-project/pull/93823 and drops the DataLayout-unaware GEP of GEP fold entirely. All cases are now left to the DataLayout-aware constant folder, which will fold everything to a single i8 GEP.

We didn't have any test coverage for this fold in LLVM, but some Clang tests change.

### CODE REVIEW

**Comment by dtcxzyw** [comment]
See https://github.com/llvm/llvm-project/commit/51e459a561519c8d51e2b4cadddc0d1f99c8b7ef:

> This causes some non-trivial text size increases in unoptimized builds for Bullet. Revert while I investigate.

Can you double check this?

**Review by aeubanks — APPROVED** [review]

**Comment by nikic** [comment]
> See [51e459a](https://github.com/llvm/llvm-project/commit/51e459a561519c8d51e2b4cadddc0d1f99c8b7ef):
> 
> > This causes some non-trivial text size increases in unoptimized builds for Bullet. Revert while I investigate.
> 
> Can you double check this?

This was mitigated by https://github.com/llvm/llvm-project/pull/93956. I did check that this patch does not cause code size regression for O0.

**Review by dtcxzyw — APPROVED** [review]
LGTM.

**Comment by jplehr** [comment]
I believe this broke our flang+openmp+offload bot: https://lab.llvm.org/staging/#/builders/140/builds/10168
Happy to help looking into it.

**Comment by nikic** [comment]
> I believe this broke our flang+openmp+offload bot: https://lab.llvm.org/staging/#/builders/140/builds/10168 Happy to help looking into it.

That's for the heads-up, I've reverted this patch in https://github.com/llvm/llvm-project/commit/cece0a105b29dcbb9d88d0aa264c4745c07a8456.

I would appreciate some help with debugging this, I don't think I can run this test locally. Probably just having the pre-optimization LLVM IR/bitcode would be enough. (Though with offload, there is probably more than one module involved?)

**Comment by jplehr** [comment]
Thank you @nikic. 
I'll see to reproduce locally and narrow down as much as possible to provide small reproducer.

**Comment by agozillon** [comment]
Hey, @jplehr notified me of this issue (thank you), I've had a little look at it and it's related to an issue this PR fixes: https://github.com/llvm/llvm-project/pull/94541 which you are currently a reviewer of ironically (as I use and alter slightly a piece of functionality you ported, so thank you very much for that)! A slightly funny convergence there :-) 

Currently when generating the kernel we try to replace uses of values with their appropriate kernel input argument, in certain cases these are Constant's which poses an issue as we can't replace them with the Input arguments which are themselves not Constant I believe. So we can only replace instructions, to do so we transform the Constants to Instructions where necessary, the current function that's in place (a small naive function from myself) will do this within the kernel to a small extent, it doesn't handle nesting's very well for example, and apparently IR this patch will produce in the test cases that are failing. However, the PR I have opened above uses the more robust and comprehensive piece of functionality you made/ported via the function `convertUsersOfConstantsToInstructions` which solves the issues encountered when this PR is landed alongside some other issues I encountered recently.

So two ways forward I can see, depending on how urgent/annoying it is for this PR to be stalled:
1.  We disable the offending tests that are breaking the buildbot until the PR that will fix them lands, which will allow you to commit this as soon as they're disabled without angering the buildbot
2. We hold off landing this until the PR that fixes the issues in the tests lands, I am unfortunately not sure how long this will take as it's been up for a while with a lack of reviewers (having a review from you @nikic would be excellent as well for the minor additions to the functionality you added), although, I can likely chase some people up to give it a look over as it's urgent. 

I don't mind which route we go down, if we opt to disable the tests, I can land a commit shortly to do so, unless either of you @nikic @jplehr would prefer to do so :-)

**Comment by nikic** [comment]
@agozillon Thanks for looking into this! This change isn't urgent, so I'm happy to wait for your PR to go in first.

**Comment by agozillon** [comment]
@nikic sounds good, if that changes don't hesitate to ping me and I'll be happy to deactivate the tests for a while (or feel free to do so yourself, just ping to tell me they're off so I know to turn them on again when the PR lands!)

**Comment by agozillon** [comment]
Just landed the PR, so it should be good to land this PR now without causing issues with the flang+openmp+offload buildbot, however, if any other issues do arise, I am happy to look into them! But from testing this PR in conjunction with the one I just landed locally, it does seem like everything should be fine.

**Pull request merged**
